// @gathertown/gather-game-common/dist/src/GameWsCloseCode.js
"use strict";
exports.__esModule = true;
exports.GameWsCloseCode = void 0;
var GameWsCloseCode;
(function (GameWsCloseCode) {
    GameWsCloseCode[GameWsCloseCode["PROTOCOL_ERROR"] = 1002] = "PROTOCOL_ERROR";
    GameWsCloseCode[GameWsCloseCode["NO_STATUS_RECEIVED"] = 1005] = "NO_STATUS_RECEIVED";
    GameWsCloseCode[GameWsCloseCode["INTERNAL_ERROR"] = 1011] = "INTERNAL_ERROR";
    GameWsCloseCode[GameWsCloseCode["SPACE_AT_CAPACITY"] = 1013] = "SPACE_AT_CAPACITY";
    GameWsCloseCode[GameWsCloseCode["CONN_TIMED_OUT"] = 4008] = "CONN_TIMED_OUT";
    GameWsCloseCode[GameWsCloseCode["SPACE_REASSIGNED"] = 4009] = "SPACE_REASSIGNED";
    GameWsCloseCode[GameWsCloseCode["CLIENT_KICKED"] = 4010] = "CLIENT_KICKED";
    GameWsCloseCode[GameWsCloseCode["UNAUTHORIZED"] = 4003] = "UNAUTHORIZED";
    GameWsCloseCode[GameWsCloseCode["CHILL_OUT"] = 4420] = "CHILL_OUT";
})(GameWsCloseCode = exports.GameWsCloseCode || (exports.GameWsCloseCode = {}));
// useful: https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent
