// @gathertown/gather-game-common/dist/src/TutorialTask.js
"use strict";
exports.__esModule = true;
exports.TutorialTaskIds = void 0;
var TutorialTaskIds;
(function (TutorialTaskIds) {
    TutorialTaskIds["MOVEMENT"] = "MOVEMENT";
    TutorialTaskIds["MUTING_UNMUTING"] = "MUTING_UNMUTING";
    TutorialTaskIds["INTERACTING"] = "INTERACTING";
    TutorialTaskIds["BETA_FEEDBACK"] = "BETA_FEEDBACK";
    // Sandbox specific
    TutorialTaskIds["SANDBOX_USING_PORTALS"] = "SANDBOX_PORTAL";
    // Office demo specific
    TutorialTaskIds["OFFICE_DEMO_REMOVING_OBJECTS"] = "OFFICE_DEMO_REMOVING_OBJECTS";
    TutorialTaskIds["OFFICE_DEMO_ADDING_OBJECTS"] = "OFFICE_DEMO_ADDING_OBJECTS";
    TutorialTaskIds["OFFICE_DEMO_DESK_CUSTOMIZATIONS"] = "OFFICE_DEMO_DESK_CUSTOMIZATIONS";
    TutorialTaskIds["OFFICE_DEMO_QUIET_MODE"] = "OFFICE_DEMO_QUIET_MODE";
    TutorialTaskIds["OFFICE_DEMO_PRIVATE_SPACES"] = "OFFICE_DEMO_PRIVATE_SPACES";
    TutorialTaskIds["OFFICE_DEMO_WHITEBOARD"] = "OFFICE_DEMO_WHITEBOARD";
    TutorialTaskIds["OFFICE_DEMO_DOCUMENTS"] = "OFFICE_DEMO_DOCUMENTS";
    TutorialTaskIds["OFFICE_DEMO_TV_DASHBOARD"] = "OFFICE_DEMO_TV_DASHBOARD";
    TutorialTaskIds["OFFICE_DEMO_SPOTLIGHT"] = "OFFICE_DEMO_SPOTLIGHT";
    TutorialTaskIds["OFFICE_DEMO_SCREENSHARE"] = "OFFICE_DEMO_SCREENSHARE";
    TutorialTaskIds["OFFICE_DEMO_EMOTES"] = "OFFICE_DEMO_EMOTES";
    TutorialTaskIds["OFFICE_DEMO_LEAVING_ROOMS"] = "OFFICE_DEMO_LEAVING_ROOMS";
    // NPC OFFICE
    TutorialTaskIds["NEW_OFFICE_DEMO_WALKING"] = "NEW_OFFICE_DEMO_WALKING";
    TutorialTaskIds["NEW_OFFICE_DEMO_DESK"] = "NEW_OFFICE_DEMO_DESK";
    TutorialTaskIds["NEW_OFFICE_DEMO_SPOTLIGHT"] = "NEW_OFFICE_DEMO_SPOTLIGHT";
    //RW Onboarding
    TutorialTaskIds["RW_OPEN_BUILD_TAB"] = "RW_OPEN_BUILD_TAB";
    TutorialTaskIds["RW_BUILD_LOBBY_OBJS"] = "RW_BUILD_LOBBY_OBJS";
    TutorialTaskIds["RW_BUILD_LOBBY_FINISH"] = "RW_BUILD_LOBBY_FINISH";
})(TutorialTaskIds = exports.TutorialTaskIds || (exports.TutorialTaskIds = {}));
